polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -28,305 +31,365 @@ struct ParallelLICM : public ParallelLICMBase<ParallelLICM> {
-  if (!llvm::all_of(op->getOperands(), definedOutside))
-    return false;
-
-  if (auto memEffect = dyn_cast<MemoryEffectOpInterface>(op)) {
-    SmallVector<MemoryEffects::EffectInstance, 1> effects;
-    memEffect.getEffects(effects);
-
-    SmallVector<SideEffects::Resource *> readResources;
-    SmallVector<SideEffects::Resource *> writeResources;
-    SmallVector<SideEffects::Resource *> freeResources;
-    for (auto effect : effects) {
-      if (isa<MemoryEffects::Allocate>(effect.getEffect()))
-        return false;
-      if (isa<MemoryEffects::Read>(effect.getEffect()))
-        readResources.push_back(effect.getResource());
-      if (isa<MemoryEffects::Write>(effect.getEffect()))
-        writeResources.push_back(effect.getResource());
-      if (isa<MemoryEffects::Free>(effect.getEffect()))
-        freeResources.push_back(effect.getResource());
+/// Represents the memory effects associated with an operation.
+struct OperationMemoryEffects {
+  OperationMemoryEffects(Operation &op) {
+    if (op.getName().hasInterface<MemoryEffectOpInterface>()) {
+      SmallVector<MemoryEffects::EffectInstance, 1> effects;
+      cast<MemoryEffectOpInterface>(op).getEffects(effects);
+
+      // Collect the memory effects of the operation.
+      for (MemoryEffects::EffectInstance &effect : effects)
+        TypeSwitch<MemoryEffects::Effect *>(effect.getEffect())
+            .Case<MemoryEffects::Read>(
+                [&](auto) { readResources.push_back(effect.getResource()); })
+            .Case<MemoryEffects::Write>(
+                [&](auto) { writeResources.push_back(effect.getResource()); })
+            .Case<MemoryEffects::Free>(
+                [&](auto) { freeResources.push_back(effect.getResource()); })
+            .Case<MemoryEffects::Allocate>([&](auto) {
+              allocateResources.push_back(effect.getResource());
+            });
     }
+  }
 
-    std::function<bool(Operation *)> conflicting = [&](Operation *b) {
-      if (willBeMoved.count(b))
-        return false;
-
-      if (b->hasTrait<OpTrait::HasRecursiveMemoryEffects>()) {
+  bool readsFromMemory() const { return !readResources.empty(); }
+  bool writesToMemory() const { return !writeResources.empty(); }
+  bool freesMemory() const { return !freeResources.empty(); }
+  bool allocatesMemory() const { return !allocateResources.empty(); }
+
+  /// Returns true if the given operation \p other has memory effects that
+  /// conflict with the memory effects summarized in this class, and false
+  /// otherwise.
+  bool conflictsWith(Operation &other) const {
+    // Check all the nested operations if 'other' has recursive side effects.
+    bool hasRecursiveEffects =
+        other.hasTrait<OpTrait::HasRecursiveMemoryEffects>();
+    if (hasRecursiveEffects) {
+      for (Region &region : other.getRegions())
+        for (Operation &innerOp : region.getOps())
+          if (conflictsWith(innerOp))
+            return true;
```

Comment:
Can we use `walkers` here?

Reply:
what is the advantage ?

Reply:
It improves readability IMO.

Reply:
Hmmm.. I find the code is reasonable, I changed the original code to avoid traversing each block explicitly